### PR TITLE
Fix/5136 node and miner

### PR DESCRIPTION
**Pull request by jcnelson**
This fixes the node and miner code paths for #5136.  It makes it so that the miner will never time out waiting for signatures, and makes it so that the block validation endpoint checks that the proposed block builds on a canonical tenure, and builds on an acceptably-high block in that tenure.

Leaving as a draft for now until integration tests pass, and until additional unit test coverage can be added.

### CODE REVIEW

**Comment by jcnelson** [comment]
I'm retargeting this to @jferrant's PR, since her PR subsumes this one anyway.  @obycode @jferrant can you take a quick look and approve?  This is a subset of @jferrant's PR.

File: stackslib/src/net/api/gettenuretip.rs
Comment (obycode) [inline stackslib/src/net/api/gettenuretip.rs:2]: Nit: shouldn't new files just have 2024 here?

**Review by obycode — APPROVED** [review]
LGTM

File: stackslib/src/net/api/gettenuretip.rs
Comment (jcnelson) [reply stackslib/src/net/api/gettenuretip.rs:121]: I don't think that matters, because the client (the Stacks signer) simply errors out if the response isn't an HTTP 200.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:488]: No, I think the only right thing to do here is for the miner to try and wait forever until:

* enough signers reject the block
* enough signers sign the block (evidenced by either the arrival of enough signer signatures via StackerDB, or the signed Nakamoto block)
* the burnchain tip changes, and `config.miner.wait_on_interim_blocks` seconds have passed since it changed

Anything else invites the possibility that the miner produces arbitrarily many siblings at the same height, since network delays alone can ensure that the signers all see a bunch of block proposals in rapid succession (as perceived by each signer), which they would all sign and return signatures for (thereby creating siblings).  This is something I thought we wanted to avoid, per the discussion in #5132.

Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:488]: Sure, that's a good idea.

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/signer/v0.rs:2069]: I'll add this back in

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (obycode) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:782]: Thinking about this some more since our last discussion... what does the miner gain from waiting for the block approval in this case?
1. It knows it cannot mine any blocks after this one in this tenure.
2. It knows that it will eventually receive the signed block via the signers.
3. It knows a new tenure has started, so it is in its best interest to submit a block commit quickly.
4. If it won the new tenure, it is in its best interest to start mining blocks, to ensure it gets its coinbase before the next tenure starts.

Based on that, I propose to drop this time, and just immediately return when it sees the new tenure.

**Comment by obycode** [comment]
I fixed `signer_set_rollover` in `develop` but the fix is not present here. That has me worried about other merge issues.

```diff
--- a/testnet/stacks-node/src/tests/signer/v0.rs
+++ b/testnet/stacks-node/src/tests/signer/v0.rs
@@ -2832,7 +2832,7 @@ fn signer_set_rollover() {
         .running_nodes
         .btc_regtest_controller
         .get_burnchain()
-        .reward_cycle_to_block_height(next_reward_cycle)
+        .nakamoto_first_block_of_cycle(next_reward_cycle)
         .saturating_add(1);
 
     info!("---- Mining to next reward set calculation -----");
```

**Comment by jcnelson** [comment]
@obycode That change is present in this PR

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:782]: Alright, I'll remove.  IIRC the rationale was that we wanted to give the signer a chance to "finish up" any in-flight blocks, but that should be irrelevant now that signers can take care of processing block signing and broadcasting themselves.

**Comment by jcnelson** [comment]
ping @kantai for a re-approval

**Comment by jcnelson** [comment]
ping @obycode for a re-approval

**Review by obycode — DISMISSED** [review]
🎉

**Comment by kantai** [comment]
This LGTM, but there are a lot of failing integration tests on this PR. I think those need to be resolved before merging.

**Review by kantai — DISMISSED** [review]
The integration tests all pass now.

There remain a handful of flaky tests that often need reruns in CI. That can be investigated separately from this PR, I think. Two tests pass locally just fine but always timeout in CI without any logging information. Those tests are disabled for now, but really should be investigated.

File: stackslib/src/net/api/postblock_proposal.rs
Comment (jferrant) [inline stackslib/src/net/api/postblock_proposal.rs:208]: This function get_highest_known_block_header_in_tenure says "Do not use in consensus code"...should the function check_block_builds_on_highest_block_in_tenure also have this warning since it calls it?

**Review by jferrant — DISMISSED** [review]

File: stackslib/src/net/api/postblock_proposal.rs
Comment (kantai) [reply stackslib/src/net/api/postblock_proposal.rs:210]: Sure -- done.

**Review by kantai — APPROVED** [review]
The integration tests all pass now.

There remain a handful of flaky tests that often need reruns in CI. That can be investigated separately from this PR, I think. Two tests pass locally just fine but always timeout in CI without any logging information. I'm pretty sure its a CI issue -- disabling those two tests causes _two other_ tests to instead exhibit the same behavior. So, it seems like it must be something weird is happening (deterministically) when it comes to executor assignment for those two tests?

**Review by jferrant — APPROVED** [review]

**Pull request merged**
